### ISSUE TITLE
Can not use slack state from reactor.

### ISSUE THREAD

**Issue by scubahub** [saltstack/salt#35716]
### Description of Issue/Question

Getting auth failure when I try to use slack state from a reactor.  Works fine from command line.

### Steps to Reproduce Issue

If I issue the following on the command line:
```salt-call slack.post_message channel="#oscar-test" from_name="ops-script" message="<removed>" api_key=<removed>```
it works as expected.  

However if I use the same arguments to slack.post_message in a reactor:
```send-failure-notification-to-slack:
  local.slack.post_message:
    - tgt: {{ salt["grains.get"]("nodename") }}  {# can not use 'id' here because it has '_master' appended to it #}
    - arg:
      - "#oscar-test"
      - message: "{{ message }}"
      - from_name: "ops-script"
      - api_key: "<removed>"
    - kwargs:
      - test: False
```
it fails.

```
2016-08-23 21:20:09,567 [salt.utils.http  ][DEBUG   ][27576] Requesting URL https://slack.com/api/chat.postMessage?token=%7B%27api_key%27%3A+%27<removed>%27%7D using POST method
2016-08-23 21:20:09,568 [salt.utils.http  ][TRACE   ][27576] Request POST Data: 'username=%7B%27from_name%27%3A+%27ops-script%27%7D&text=%7B%27message%27%3A+%27<removed>%27%7D&channel=%23oscar-test'
2016-08-23 21:20:10,010 [salt.utils.http  ][DEBUG   ][27576] Response Status Code: 200
2016-08-23 21:20:10,011 [salt.utils.http  ][TRACE   ][27576] Response Headers: {'X-Consumed-Content-Encoding': 'gzip', 'Content-Length': '55', 'X-Xss-Protection': '0', 'X-Content-Type-Options': 'nosniff', 'Content-Security-Policy': 'referrer no-referrer;', 'X-Cache': 'Miss from cloudfront', 'X-Slack-Req-Id': '26d46462-4685-47a4-8701-a6d1c9ea3919', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload', 'Vary': 'Accept-Encoding', 'Server': 'Apache', 'Connection': 'close', 'X-Slack-Backend': 'h', 'X-Amz-Cf-Id': '<removed>', 'Via': '1.1 199c9bce22bd411402daf00db8dbb17d.cloudfront.net (CloudFront)', 'Date': 'Tue, 23 Aug 2016 21:20:09 GMT', 'Access-Control-Allow-Origin': '*', 'Content-Type': 'application/json; charset=utf-8'}
2016-08-23 21:20:10,011 [salt.utils.http  ][TRACE   ][27576] Response Cookies: None
2016-08-23 21:20:10,011 [salt.utils.http  ][TRACE   ][27576] Response Text: {"ok":false,"error":"invalid_auth"}
```

For comparison, here are the logs when I use salt-call with the same arguments:
```
2016-08-23 21:25:15,350 [salt.utils.http  ][DEBUG   ][29236] Requesting URL https://slack.com/api/chat.postMessage?token=<removed> using POST method
2016-08-23 21:25:15,350 [salt.utils.http  ][TRACE   ][29236] Request POST Data: 'username=ops-script&text=<removed>&channel=%23oscar-test'
2016-08-23 21:25:15,820 [salt.utils.http  ][DEBUG   ][29236] Response Status Code: 200
2016-08-23 21:25:15,820 [salt.utils.http  ][TRACE   ][29236] Response Headers: {'X-Consumed-Content-Encoding': 'gzip', 'Content-Length': '249', 'Content-Security-Policy': 'referrer no-referrer;', 'Access-Control-Allow-Origin': '*', 'X-Accepted-Oauth-Scopes': 'chat:write:bot,post', 'Expires': 'Mon, 26 Jul 1997 05:00:00 GMT', 'Pragma': 'no-cache', 'Date': 'Tue, 23 Aug 2016 21:25:15 GMT', 'X-Xss-Protection': '0', 'X-Cache': 'Miss from cloudfront', 'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload', 'Server': 'Apache', 'Connection': 'close', 'X-Amz-Cf-Id': '<removed>', 'X-Slack-Req-Id': '70466d2b-9e19-4df3-b805-b6d84dae1c42', 'Via': '1.1 199c9bce22bd411402daf00db8dbb17d.cloudfront.net (CloudFront)', 'X-Content-Type-Options': 'nosniff', 'Vary': 'Accept-Encoding', 'X-Slack-Backend': 'h', 'Cache-Control': 'private, no-cache, no-store, must-revalidate', 'X-Oauth-Scopes': 'identify,read,post,client,apps', 'Content-Type': 'application/json; charset=utf-8'}
2016-08-23 21:25:15,820 [salt.utils.http  ][TRACE   ][29236] Response Cookies: None
2016-08-23 21:25:15,820 [salt.utils.http  ][TRACE   ][29236] Response Text: {"ok":true,"channel":"G2361AAE9","ts":"1471987515.000005","message":{"text":"<removed>","username":"ops-script","bot_id":"<removed>","type":"message","subtype":"bot_message","ts":"1471987515.000005"}}
```
### Versions Report

```
Salt Version:
           Salt: 2015.8.11

Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: 0.21.1
           Mako: Not Installed
         PyYAML: 3.11
          PyZMQ: 14.7.0
         Python: 2.7.5 (default, Nov 20 2015, 02:00:19)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.5
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.7
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: Not Installed
        timelib: Not Installed

System Versions:
           dist: centos 7.2.1511 Core
        machine: x86_64
        release: 3.10.0-229.14.1.el7.x86_64
         system: CentOS Linux 7.2.1511 Core
```

**Comment by jbouse** [comment]
Just an initial gut check first glance but have you tried removing the double quotes (") from all the items except around the channel and moving them to the kwargs?

For example
```
  local.slack.post_message:
    - tgt: {{ salt["grains.get"]("nodename") }}  {# can not use 'id' here because it has '_master' appended to it #}
    - arg:
      - '#oscar-test'
    - kwargs:
      - test: False
      - message: {{ message }}
      - from_name: ops-script
      - api_key: <removed>
```

**Comment by scubahub** [comment]
That does not work:
```
2016-08-24 16:45:12,384 [salt.minion      ][WARNING ][19122] Passed invalid arguments to slack.post_message: post_message() takes at least 3 arguments (1 given)

    Send a message to a Slack channel.

    :param channel:     The channel name, either will work.
    :param message:     The message to send to the Slack channel.
    :param from_name:   Specify who the message is from.
    :param api_key:     The Slack api key, if not specified in the configuration.
    :return:            Boolean if message was sent successfully.

    CLI Example:

    .. code-block:: bash

        salt '*' slack.post_message channel="Development Room" message="Build is done" from_name="Build Server"


Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/minion.py", line 1078, in _thread_return
    return_data = func(*args, **kwargs)
TypeError: post_message() takes at least 3 arguments (1 given)
```

**Comment by jbouse** [comment]
Okay that was just an off the cuff suggestion... I've not used the slack.post_message myself yet but I do have Slack configured as a returner and engine. I'll try to play around with it later.

**Comment by scubahub** [comment]
Turns out I missed a key part (pun intended) which was in the part of the logs I snipped out.  The Requesting URL log line shows the encoding of the api_key is different.  When I use salt-call the url contains ```?token=$API_KEY``` but when I invoke it from the reactor it contains ```?token=%7B%27api_key%27%3A+%27$API_KEY%27%7D```.  So I took out all the variable names:

```
send-failure-notification-to-slack:
  local.slack.post_message:
    - tgt: {{ salt["grains.get"]("nodename") }}  {# can not use 'id' here because it has '_master' appended to it #}
    - arg:
      - "#oscar-test"
      - {{ message }}
      - ops-script
      - $API_KEY
    - kwargs:
      - test: False
```

And now it works.

Looks like this is a non issue, but rather due to me just starting to use reactor.  Reactor invokes execution modules not state modules and I was looking at https://docs.saltstack.com/en/2015.8/ref/states/all/salt.states.slack.html#salt.states.slack.post_message rather than https://docs.saltstack.com/en/2015.8/ref/modules/all/salt.modules.slack_notify.html#salt.modules.slack_notify.post_message when trying to figure out how to write the reactor.
